sourced/ml/transformers/coocc.py

```diff
@@ -0,0 +1,87 @@
+import itertools
+import operator
+
+from scipy import sparse
+from bblfsh import role_id
+
+from sourced.ml.models import Cooccurrences
+from sourced.ml.transformers import Transformer
+
+
+class CooccModelSaver(Transformer):
+    def __init__(self, output, tokens_list, **kwargs):
+        super().__init__(**kwargs)
+        self.tokens_list = tokens_list
+        self.output = output
+
+    def __call__(self, sparce_matrix):
+        matrix_count = sparce_matrix.count()
+        rows = sparce_matrix.take(matrix_count)
+
+        mat_row, mat_col, mat_weights = zip(*rows)
+        tokens_num = len(self.tokens_list)
+        matrix = sparse.coo_matrix((mat_weights, (mat_row, mat_col)),
+                                   shape=(tokens_num, tokens_num))
+
+        Cooccurrences().construct(self.tokens_list, matrix).save(self.output)
+
+
+class CooccConstructor(Transformer):
+    def __init__(self, token2index, token_parser, prune_size=1, **kwargs):
+        super().__init__(**kwargs)
+        self.token2index = token2index
+        self.token_parser = token_parser
+
+        # TODO(zurk): implement pruning
+        self.prune_size = prune_size
+
+    def _flatten_children(self, root):
+        ids = []
+        stack = list(root.children)
+        for node in stack:
+            if role_id("IDENTIFIER") in node.roles and role_id("QUALIFIED") not in node.roles:
```

Comment:
I am crying. Please, `bblfsh.filter`

Reply:
I can not. I need to check only one Node. We need a new feature. Not to filter, but to check `Node` for some condition.

Reply:
come on. This code traverses the tree and picks the ids. In a really bad and an invalid way BTW, because it is undefined if you change the list which you iterate.
```
return list(bblfsh.filter(root, "@//whatever"))
```

Reply:
Do you think it must filter only the direct children? Then we must punish the person who wrote it.